Sources/NIOHTTP1Client/main.swift

```diff
@@ -0,0 +1,100 @@
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOHTTP1
+import NIOOpenSSL
+import Foundation
+
+private final class HTTPResponseHandler: ChannelInboundHandler {
+
+    let promise: EventLoopPromise<Void>
+
+    init(_ promise: EventLoopPromise<Void>) {
+        self.promise = promise
+    }
+
+    typealias InboundIn = HTTPClientResponsePart
+
+    func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
+        let httpResponsePart = unwrapInboundIn(data)
+        switch httpResponsePart {
+        case .head(let httpResponseHeader):
+            print("\(httpResponseHeader.version) \(httpResponseHeader.status.code) \(httpResponseHeader.status.reasonPhrase)")
+            for (name, value) in httpResponseHeader.headers {
+                print("\(name): \(value)")
+            }
+        case .body(var byteBuffer):
+            if let responseBody = byteBuffer.readString(length: byteBuffer.readableBytes) {
+                print(responseBody)
+            }
+        case .end(_):
+            _ = ctx.channel.close()
+            promise.succeed(result: ())
+        }
+    }
+
+    func errorCaught(ctx: ChannelHandlerContext, error: Error) {
+        print("Error: ", error)
+        _ = ctx.channel.close()
+        promise.succeed(result: ())
+    }
+}
+
+let arguments = CommandLine.arguments
+let arg1 = arguments.dropFirst().first
+
+var url: URL
+
+if let u = arg1 {
+    url = URL(string: u)!
+} else {
+    url = URL(string: "https://::1:4433/get")!
+}
+
+let eventLoopGroup = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+let promise: EventLoopPromise<Void> = eventLoopGroup.next().newPromise()
+defer {
+    try! promise.futureResult.wait()
+    try! eventLoopGroup.syncShutdownGracefully()
+}
+
+let tlsConfiguration = TLSConfiguration.forClient()
+let sslContext = try! SSLContext(configuration: tlsConfiguration)
+
+let bootstrap = ClientBootstrap(group: eventLoopGroup)
+        .channelOption(ChannelOptions.socket(SocketOptionLevel(SOL_SOCKET), SO_REUSEADDR), value: 1)
+        .channelInitializer { channel in
+            let openSslHandler = try! OpenSSLClientHandler(context: sslContext, serverHostname: url.host)
+            return channel.pipeline.add(handler: openSslHandler).then {
+                channel.pipeline.addHTTPClientHandlers()
+            }.then {
+                channel.pipeline.add(handler: HTTPResponseHandler(promise))
+            }
+        }
+
+func sendRequest(_ channel: Channel) -> EventLoopFuture<Void> {
+    var request = HTTPRequestHead(version: HTTPVersion(major: 1, minor: 1), method: HTTPMethod.GET, uri: url.absoluteString)
+    request.headers = HTTPHeaders([
+        ("Host", url.host!),
+        ("User-Agent", "swift-nio"),
+        ("Accept", "application/json")
```

Comment:
I think you should add `("Connection", "close")` as you're closing the connection after receiving the response.

Reply:
Right. But if I add this header, `HTTPResponseDecoder` signals `fireChannelInactive` earlier than `fireChannelRead` even though it has pending inputs.

Reply:
I changed in 10c09d5 to use `Connection` header, but currently it seems that this handler's `channelRead` is not called.

Reply:
Are you sure? Please try updating to the current master, because when I run this patch against current master it seems that channelRead is called fine.

Reply:
Aha, the issue appears to be intermittent.

Reply:
Ok, there's a straightforward logic bug in this module. I'll fix it up.

Reply:
awesome job @mike-neck and @Lukasa for finding this issue and fixing it!

Reply:
@Lukasa Thanks it's working fine!